### ISSUE TITLE
Very slow to generate coverage report

### ISSUE THREAD

**Issue by janaka** [tonerdo/coverlet#145]
Opening a new issue to track my problem we were discussing in #138 (you need to add the package ref).

This is the repo I've added Coverlet to https://github.com/ivansams/CoverletCheck. There's only one test but it takes 5+ minutes to output the coverage report in the terminal and say it has generated the file. On macOS. Seems to also make the process run at 100% CPU.

**Comment by tonerdo** [comment]
Switch to version 2.1.0. It includes massive performance improvements

**Comment by janaka** [comment]
.NET Core 2.1?

**Comment by tonerdo** [comment]
No I mean coverlet: https://www.nuget.org/packages/coverlet.msbuild/2.1.0

**Comment by janaka** [comment]
Switched to coverlet 2.1.0. Still very very very slow, multiple minutes.

**Comment by tonerdo** [comment]
Hmm, that's weird. How large is your project?

**Comment by petli** [comment]
But I do notice it takes a few seconds between saying "Generating report..." and outputting the coverage stats.  For such a small test that should not be noticable, so there might be something there that is much exarcerbated on MacOS.

**Comment by petli** [comment]
@janaka Can you try the opencover output format? That seems to be much quicker than the default coverlet JSON format.

    -p:CoverletOutputFormat=opencover

**Comment by tonerdo** [comment]
Hmm... @petli perhaps it's the speed of the serialization library?

**Comment by petli** [comment]
@tonerdo Newtonsoft.Json is the most commonly used JSON library (maybe even now part of standard .NET?) and it tend to be fast.

**Comment by tonerdo** [comment]
Yeah, I used Jil because my Newtonsoft.Json dependency was conflicting with that of MSBuild and I didn't have much time to look into it. Will change it before the next release

**Comment by tonerdo** [comment]
@janaka I cloned the project on my Mac and the json result only takes `4s` to generate which leads me to explore the possibility of a problem specific to your machine. Either way I've changed the json library to `Newtonsoft` and report generation reduced to `0.09s`. The release with this change should go out Friday

**Comment by janaka** [comment]
Hmm interesting. Definitely took minutes on my MBP (actually I think I happened to try on both my 2015 MBP and 2017 MBP). When I try this next I'll push the exact repo that I'm testing with.

**Comment by joaosilvareed** [comment]
This also happens to me with this example.

**Comment by turbo** [comment]
I too am experiencing a massive performance impact for instrumented runs. One of my test projects is launched like this:

```bash
echo "" | dotnet run -p TCLS ALL
```

It runs about 250,000 tests and completes in about 25 to 30 minutes. However, when run with coverlet:

```bash
coverlet \
    "TCLS/bin/Debug/netcoreapp2.1/TCLS.dll" \
    --target "./tcltests" \
    --targetargs "run --no-build -p TCLS ALL" \
    --exclude "[xunit.*]*" \
    --exclude "[*.uninstrumented]" \
    --exclude "[*.Core]SupportClass*" \
    --exclude "[*.Core]Testing*" \
    --output "cov-results/TCLS/" \
    -f json -f opencover
```

- it takes many minutes before the test program is even launched
- it takes 100 minutes to complete (though coverlet accurately collects metrics)

The first is also true for any other projects I've tried with coverlet, even just with one test in it. It takes forever to even start and then runs *very* slow. Consistently about 30% performance overhead.

Tested this using .NET core on Linux and Windows:

- .NET Core SDK 2.1
- Instrumented assemblies all .netstandard 2.0 or .netcoreapp2.1

Omitting the json option also doesn't speed up the test runs.

**Comment by tonerdo** [comment]
@turbo what version of Coverlet are you using?

**Comment by turbo** [comment]
I've used the global tool installer command to install it, this is the version output:

```bash
$ coverlet --version
Cross platform .NET Core code coverage tool
1.4.0.0
```

**Comment by turbo** [comment]
Running everything in release mode brings the total test execution down to 60 minutes overall. One bottleneck is coverlet not being run in parallel. Right now, I have 5 different tests assemblies and coverlet merged the results of all of them. However, that means I need to run one after the other. If I could run many coverlet instances at once and later *just* merge the results into one result file, I'd see a pretty significant reduction in overall test time.
